src/shared/codelens/localLambdaRunner.ts

```diff
@@ -0,0 +1,335 @@
+import { SamCliProcessInvoker, SamCliTaskInvoker } from '../sam/cli/samCliInvoker'
+import { SamCliLocalInvokeInvocation } from '../sam/cli/samCliLocalInvoke'
+import { SettingsConfiguration } from '../settingsConfiguration'
+import { SamTemplateGenerator } from '../templates/sam/samTemplateGenerator'
+import { ExtensionDisposableFiles } from '../utilities/disposableFiles'
+
+import { getLogger, OutputChannelName } from './codeLensUtils'
+
+const localize = nls.loadMessageBundle()
+
+const logger = getLogger(OutputChannelName.Lambda)
+
+export interface LambdaLocalInvokeArguments {
+  document: vscode.TextDocument,
+  range: vscode.Range,
+  handlerName: string,
+  debug: boolean,
+  workspaceFolder: vscode.WorkspaceFolder
+}
+
+export interface SAMTemplateEnvironmentVariables {
+  [resource: string]: {
+      [key: string]: string
+  }
+}
+
+export class LocalLambdaRunner {
+
+  private static readonly TEMPLATE_RESOURCE_NAME: string = 'awsToolkitSamLocalResource'
+  private static readonly SAM_LOCAL_PORT_CHECK_RETRY_INTERVAL_MILLIS: number = 125
+  private static readonly SAM_LOCAL_PORT_CHECK_RETRY_TIMEOUT_MILLIS_DEFAULT: number = 30000
+
+  private _baseBuildFolder?: string
+  private readonly _debugPort?: number
+
+  public constructor(
+      private readonly configuration: SettingsConfiguration,
+      private readonly localInvokeArgs: LambdaLocalInvokeArguments,
+      debugPort: number | undefined,
+      private readonly runtime: string,
+      private readonly outputChannel: vscode.OutputChannel,
+      private readonly processInvoker: SamCliProcessInvoker,
+      private readonly taskInvoker: SamCliTaskInvoker,
+      private readonly debugConfig: vscode.DebugConfiguration,
+      private readonly codeRootDirectoryPath: string,
+      private readonly onWillAttachDebugger?: () => Promise<void>,
+      private readonly onDidSamBuild?: () => Promise<void>
+  ) {
+      if (localInvokeArgs.debug && !debugPort) {
+          throw new Error('Debug port must be provided when launching in debug mode')
+      }
+
+      this._debugPort = debugPort
+  }
+
+  public async run(): Promise<void> {
+      try {
+          logger.info(
+              localize(
+                  'AWS.output.sam.local.start',
+                  'Preparing to run {0} locally...',
+                  this.localInvokeArgs.handlerName
+              )
+          )
+
+          // const rootCodeFolder: string = await this.determineRootCodeFolder()
+
+          const inputTemplate: string = await this.generateInputTemplate(this.codeRootDirectoryPath)
+          const samBuildTemplate: string = await this.executeSamBuild(this.codeRootDirectoryPath, inputTemplate)
+
+          await this.invokeLambdaFunction(samBuildTemplate)
+
+      } catch (err) {
+          // TODO: logger.error?
+          console.log(err)
+          const error = err as Error
+
+        // TODO: Define standard/strategy. Sometimes err.message is/isn't part of msg "Error: {0}"
```

Comment:
Does `tslint` fail to warn you about indentation issues? The following rule (from our `tslint.yaml`) should catch things like this:

```
  indent:
    options:
      - spaces
      - 4
```

Reply:
No. Oddly it beats me over the head about all sorts of "pretty" distractions on a keystroke by keystroke basis, but not that. It would be more productive to have an auto-save and/or build time prettifier: https://github.com/aws/aws-toolkit-vscode/issues/324